mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,964 @@
+    unsigned loopNestSize) {
+
+  unsigned numOps = loadAndStoreOps.size();
+  // Resize the `loopCarriedDependenceVector` to loop nest size.
+  this->loopCarriedDependenceVector.resize(loopNestSize);
+
+  for (unsigned i = 0; i < numOps; ++i) {
+    Operation *srcOp = loadAndStoreOps[i];
+    for (unsigned j = 0; j < numOps; ++j) {
+      Operation *dstOp = loadAndStoreOps[j];
+      for (unsigned depth = 1; depth <= loopNestSize + 1; ++depth) {
+        MemRefAccess srcAccess(srcOp);
+        MemRefAccess dstAccess(dstOp);
+        FlatAffineConstraints dependenceConstraints;
+        SmallVector<DependenceComponent, 2> depComps;
+        DependenceResult result = checkMemrefAccessDependence(
+            srcAccess, dstAccess, depth, &dependenceConstraints, &depComps);
+
+        if (hasDependence(result)) {
+          for (unsigned i = 0; i < depComps.size(); i++) {
+            DependenceComponent depComp = depComps[i];
+            if (depComp.lb.getValue() != 0 || depComp.ub.getValue() != 0)
+              this->loopCarriedDependenceVector[i] = true;
+          }
+          // A dependence is found, hence no need to check on further depths.
+          break;
+        }
+      }
+    }
+  }
+}
+
+/// Calculates a representative cost of a permutation regarding parallelism on
+/// multicores. A permutation having more free outer loops gets a smaller cost.
+uint64_t getParallelismCost(ArrayRef<unsigned> perm,
+                            SmallVector<bool, 4> &loopCarriedDV,
+                            SmallVector<unsigned, 4> &iterCounts) {
+  unsigned totalcost = 0;
+  unsigned thisLoopcost = 1;
+
+  for (unsigned i = 0; i < perm.size(); i++) {
+    if (!loopCarriedDV[perm[i]])
+      continue;
+
+    thisLoopcost = 1;
+    for (unsigned j = i + 1; j < perm.size(); j++) 
+      thisLoopcost *= iterCounts[perm[j]];
+
+    totalcost += thisLoopcost;
+  }
+  return totalcost;
+}
+
+/// Removes `dstOp` op from its current group and inserts it into `srcOp` op's
+/// group. Updates `groupId` to reflect the changes.
+static void insertIntoReferenceGroup(
+    SmallVector<llvm::SmallSet<Operation *, 8>, 8> &referenceGroups,
+    DenseMap<Operation *, unsigned> &groupId, Operation *srcOp,
+    Operation *dstOp) {
+  referenceGroups[groupId[srcOp]].insert(
+      referenceGroups[groupId[dstOp]].begin(),
+      referenceGroups[groupId[dstOp]].end());
+  referenceGroups.erase(referenceGroups.begin() + groupId[dstOp]);
+  // Insert operation results in same group-id for both instructions.
+  groupId[dstOp] = groupId[srcOp];
+}
+
+/// Groups load/store ops in `loadAndStoreOps` list as reference groups to 
+/// calculate group-reuse. Two ops are in the same reference group if they 
+/// exibit group-temporal or group-spatial reuse with respect to an 
+/// affine.for loop at index `innermostForOp` in original loop nest.
+///
+/// Please refer Steve Carr et. al for a detailed description.
+/// https://dl.acm.org/doi/abs/10.1145/195470.195557
```

Comment:
Clear explain which parameters are the input and output of the function.

in original -> in the original

Reply:
Done.